add GPU support to symPACK

Hooninator opened the pull request:
This PR adds CUDA device functionality to symPACK. Large enough computations are offloaded to the GPU for faster execution. All CUDA device support/functionality is wrapped in an `#ifdef` conditional on a cmake variable `CUDA_MODE` that can be set during cmake configuration via `-DENABLE_CUDA={ON|OFF}`, so this PR does not alter the CPU mode functionality of symPACK in any way. I've compiled both versions of symPACK successfully on Perlmutter (the CPU mode was compiled with all CUDA related modules unloaded).

The files with the most substantial changes are `symPACKMatrix2D.hpp`, `cuBLAS.{cpp, hpp}`, and `global.cpp`. Some newly added command line arguments related to the GPU mode are in `utils.hpp`. 

Documentation for the new GPU functionality is forthcoming.

mathias-cerebras commented on include/sympack/symPACKMatrix2D.hpp:
```diff
@@ -365,25 +384,42 @@ namespace symPACK{
+#ifdef CUDA_MODE
+            if (is_gpu_block) {
+              upcxx::when_all(
+                upcxx::rget(remote_gptr, landing_zone, size),
+                upcxx::copy(remote_gptr, d_landing_zone, size)                    
```
Does this mean the data lands on the host first, and then is copied to the device?

Hooninator replied:
The data that's being copied here is a bucket of bytes that contains two things, a linked list of `block_t` objects, and an array containing the nonzeros of the block. The nonzeros need to be on the device, but the `block_t` objects do not. Therefore, the data is copied to both the host and the device here so each entity has access to what it needs from the bucket of bytes. 
An obvious optimization that I overlooked is to copy just the `block_t` linked list to the host and copy just the nonzeros to the device. This is definitely what I should have done, and I may or may not change it before Friday evening.

mathias-cerebras commented on include/sympack/symPACKMatrix2D.hpp:
```diff
@@ -663,8 +739,9 @@ namespace symPACK{
+
           _block_container._nblocks = other._block_container._nblocks;
-          _nnz = other._nnz;
+          _nnz = other._nnz;        
```
trailing whitespaces

mathias-cerebras commented on include/sympack/symPACKMatrix2D.hpp:
```diff
@@ -1052,11 +1374,12 @@ namespace symPACK{
                       break;
-                    }
-                  } while ( ++tgt_ptr<_block_container._blocks + _block_container._nblocks ); 
+                    }                 
```
nit: trailing whitespaces

mathias-cerebras commented on include/sympack/symPACKMatrix2D.hpp:
```diff
@@ -1052,11 +1374,12 @@ namespace symPACK{
-                    }
-                  } while ( ++tgt_ptr<_block_container._blocks + _block_container._nblocks ); 
+                    }                 
+                  } while ( ++tgt_ptr<_block_container._blocks + _block_container._nblocks );                  
```
nit: trailing whitespaces

bonachea commented:
@Hooninator thanks for all the great work on this!

You mentioned that documentation for GPU functionality is forthcoming - can you give us some idea what you're planning here?

Things I'd like the docs to include:

1. User-facing documentation : (probably in README.md)
    a. How to build/run using the new CUDA features
    b. Updated UPC++ version requirements for ENABLE_CUDA mode (at least 2023.3.0 based on source inspection)
3. Maintainer-facing documentation: should probably go in a new markdown text file at the top level (TODO.md?) to avoid confusing end users
    a. list of any known defects or limitations you encountered
    b. list of possible future enhancements that seem most likely to yield improved behavior

 Given the compressed timeline, I think it probably makes sense to continue pushing documentation updates to this PR concurrent with review.

bonachea commented on include/sympack/symPACKMatrix2D.hpp:
```diff
@@ -865,30 +958,139 @@ namespace symPACK{
+	bool gpu_alloc_check(std::vector<size_t> bytes) {
+		
+		/* Don't do it if GPU hasn't enough space */
+		size_t free = symPACK::gpu_allocator.segment_size();
```
This query doesn't do what you seem to think it does. 

`device_allocator::segment_size()` returns the TOTAL size of the segment (ie the size specified at allocator construction) and does not reflect current allocation status.

If you want an estimate of the remaining free space, you need something more like this:

```suggestion
		size_t free = symPACK::gpu_allocator.segment_size() - 
		                    symPACK::gpu_allocator.segment_used();
```

Hooninator replied:
See comment below. Neither function is actually used, so I will likely just remove both from the PR.

bonachea commented on include/sympack/symPACKMatrix2D.hpp:
```diff
@@ -865,30 +958,139 @@ namespace symPACK{
+			<<" bytes on device with "<<symPACK::gpu_allocator.segment_size()<<" space left"<<std::endl;
+		gdb_lock();
+	}
+#endif
```
The two functions above `gpu_alloc_check` and `gpu_alloc_err_handler` use library calls introduced in UPC++ 2023.3.0, and these _might_ be the only use of features introduced in that release. However these functions do not themselves appear to currently have any callers.

It's possible that adding appropriate `#if UPCXX_SPEC_VERSION >= 20230300` here could allow this code to build with older versions of UPC++

Hooninator replied:
Yeah neither function is actually used, and they should both probably be removed from the PR. The "GPU error handler" that the code currently uses is really just "do it on the CPU if the GPU allocation failed due to insufficient memory", which does not require a separate function to implement.

bonachea replied:
With those functions removed I suspect the UPC++ floor for ENABLE_CUDA will be 2022.3.0 (for `make_gpu_allocator`), although that's only based on a quick skim and it's possible I missed something.

Hooninator commented:
> @Hooninator thanks for all the great work on this!
> 
> You mentioned that documentation for GPU functionality is forthcoming - can you give us some idea what you're planning here?
> 
> Things I'd like the docs to include:
> 
> 1. User-facing documentation : (probably in README.md)
>    a. How to build/run using the new CUDA features
>    b. Updated UPC++ version requirements for ENABLE_CUDA mode (at least 2023.3.0 based on source inspection)
> 2. Maintainer-facing documentation: should probably go in a new markdown text file at the top level (TODO.md?) to avoid confusing end users
>    a. list of any known defects or limitations you encountered
>    b. list of possible future enhancements that seem most likely to yield improved behavior
> 
> Given the compressed timeline, I think it probably makes sense to continue pushing documentation updates to this PR concurrent with review.

What you requested is largely what I had in mind for the documentation. The user facing documentation will include descriptions of the cmake build variable used to activate the CUDA mode, as well as descriptions of the GPU-mode command line args. I will write up a draft of the documentation and push it to this PR.

Hooninator commented:
First pass of some documentation is added and I have removed the methods Dan mentioned in an earlier comment.

bonachea commented:
> First pass of some documentation is added and I have removed the methods Dan mentioned in an earlier comment.

@Hooninator commit [202aff8](https://github.com/symPACK/symPACK/pull/12/commits/202aff81a4c5cdfc0247e8a1f58d70ceb934edf0) removed alot more code than just the two "dead" methods discussed earlier, and the removal of `cublas` calls was not mentioned in the commit message.  

Please confirm this was intentional and not an accidental removal?

bonachea commented on src/global.cpp:
```diff
@@ -36,27 +36,82 @@ int symPACK_Rank(int * rank){
+    abort();
+  }
+
+  MPI_Comm shmcomm;
+  int tasks_per_node;
+  MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_SHARED, 0, MPI_INFO_NULL, &shmcomm);
+  MPI_Comm_size(shmcomm, &tasks_per_node);
```
The UPC++ idiom for this:

```suggestion
  int tasks_per_node = upcxx::local_team().rank_n();
```

Hooninator commented:
> > First pass of some documentation is added and I have removed the methods Dan mentioned in an earlier comment.
> 
> @Hooninator commit [202aff8](https://github.com/symPACK/symPACK/pull/12/commits/202aff81a4c5cdfc0247e8a1f58d70ceb934edf0) removed alot more code than just the two "dead" methods discussed earlier, and the removal of `cublas` calls was not mentioned in the commit message.
> 
> Please confirm this was intentional and not an accidental removal?

This was intentional. The removed cublas calls were ones that I added very early on in development, and they were not used in the standard Cholesky factorization mode of symPACK, only for the LDL^t factorization mode, which I did not fully implement GPU support for. Due to this incomplete implementation, if someone built symPACK with the CUDA mode enabled and attempted to use LDL^t factorization, it would likely either explode or produce incorrect results, so I removed the cublas calls to avoid the risk of this happening.

bonachea commented on src/global.cpp:
```diff
@@ -36,27 +36,82 @@ int symPACK_Rank(int * rank){
+  
+  size_t alloc_size, free, total;
+  CUDA_ERROR_CHECK(cudaMemGetInfo(&free, &total));
+  alloc_size = (free/2) / (std::max(tasks_per_node, n_gpus) / n_gpus);
```
This algorithm assumes that `n_gpus == cuda_device::device_n()` reflects the _total_ number of physical GPUs plugged into this compute node, which is often false.

SLURM GPU allocation options can and often do cause CUDA_VISIBLE_DEVICES to be set in a way such that `libcuda` (and hence also `cuda_device`) for each task only sees the fraction of physical GPUs assigned to that task. For example this command:
```
srun --gpus-per-node=4 -N 1 -n 4 a.out
```
results in 4 processes sharing a node that each "see" all **four** shared GPUs (as reported by `cuda_device::device_n()` and `cuDeviceGetCount()`), and the algorithm should work as intended.
However this command:
```
srun --gpus-per-task=1 -N 1 -n 4 a.out   
```
results in 4 processes sharing a node that each only see their **single** dedicated GPU, and this algorithm will underestimate the space available to each by a factor of four.
Similarly this command:
```
srun --ntasks-per-gpu=16 -N 1 -n 64 a.out
```
results in 64 processes sharing a node that each only see the **single** GPU shared with 15 peers, and this algorithm will also underestimate the space available to each by a factor of 4.

TL:DR: On systems with more than 1 physical GPU per node, some job layouts can cause this defaulting algorithm to choose a size considerably smaller than an even partitioning of the shared GPU memory resources. Please add a comment noting this deficiency and update README to explain the defaulting for `-gpu_mem` might not work optimally for systems with more than 1 GPU/node, where manual setting should be recommended.

bonachea replied:
Thanks for adding documentation in the README.

I still think a comment in the code (and/or an enhancement item in TODO.md) is also warranted to record the fact this defaulting algorithm is known to greatly underestimate the usable space for some layouts on multi-GPU systems.

bonachea commented on README.md:
The `run_sympack2D` executable is currently not mentioned anywhere in the README. Should it be? 

The PAW-ATM AD instructs readers to use `run_sympack2D`, so it would be nice to address this somehow. Are both drivers believed to be equally correct? 

Perhaps we can at least briefly describe each driver and ideally give a brief recommendation on how to choose?

Hooninator replied:
Yeah this sounds like a good idea. I will add this.

bonachea commented on README.md:
```diff
@@ -125,6 +133,15 @@ Note that to run **symPACK**, `upcxx-run` is used in the example above, but on s
 other launchers may be used to both spawn **MPI** and **UPC++**, such as `srun` if the system is using SLURM.
 Moreover, for larger problems, the `-shared-heap XX` argument to `upcxx-run` may be needed (Please refer to **UPC++** documentation).
```
Suggested link:

```suggestion
Moreover, for larger problems, the `-shared-heap XX` argument to `upcxx-run` may be needed (Please refer to [**UPC++** documentation](https://upcxx.lbl.gov/site)).
```
Please also replace the mention of NERSC Cori above with NERSC Perlmutter

bonachea commented on include/sympack/Environment.hpp:
```diff
@@ -137,7 +151,34 @@ namespace symPACK{
 
 
-
+#define CUBLAS_ERROR_CHECK(s)                                                                       \
+{                                                                                                   \
+    cublasStatus_t err;                                                                             \
+    if ((err = (s)) != CUBLAS_STATUS_SUCCESS)                                                       \
+    {                                                                                               \
+        std::cout << "cuBLAS Error " << err << " at " << __FILE__ << ":" << __LINE__ << "\n";       \
+        exit(1);                                                                                    \
+    }                                                                                               \
+}                                                                                                   \
+
+#define CUDA_ERROR_CHECK(s)                                                                         \
+{                                                                                                   \
+    cudaError_t error = s;                                                                          \
+    if (error != cudaSuccess) {                                                                     \
+        std::cout << "CUDA Error " << error << " at " << __FILE__ << ":" << __LINE__ << "\n";       \
+        std::cout << cudaGetErrorString(error) << "\n";                                             \
+        exit(1);                                                                                    \
+    }                                                                                               \
+}   
+
+#define CUSOLVER_ERROR_CHECK(s)                                                                          \
+{                                                                                                        \
+    cusolverStatus_t error = s;                                                                          \
+    if (error != CUSOLVER_STATUS_SUCCESS) {                                                              \
+        std::cout << "cuSOLVER Error " << error << " at " << __FILE__ << ":" << __LINE__ << "\n";        \
+        exit(1);                                                                                         \
+    }                                                                                                    \
+}                                                                                                        \
```
For block-statement macros like this, it's best-practice to wrap the block definition inside a `do { ... } while(0)` statement. This idiom ensures non-surprising behavior with invocations of the form:
```
if (something()) CUDA_ERROR_CHECK(some_cuda_call());
else             CUDA_ERROR_CHECK(some_other_cuda_call());
```
which would otherwise generate a parse error as the macros are currently defined.

```suggestion
#define CUBLAS_ERROR_CHECK(s)                                                                       \
do {                                                                                                   \
    cublasStatus_t err;                                                                             \
    if ((err = (s)) != CUBLAS_STATUS_SUCCESS)                                                       \
    {                                                                                               \
        std::cout << "cuBLAS Error " << err << " at " << __FILE__ << ":" << __LINE__ << "\n";       \
        exit(1);                                                                                    \
    }                                                                                               \
} while (0)

#define CUDA_ERROR_CHECK(s)                                                                         \
do {                                                                                                   \
    cudaError_t error = s;                                                                          \
    if (error != cudaSuccess) {                                                                     \
        std::cout << "CUDA Error " << error << " at " << __FILE__ << ":" << __LINE__ << "\n";       \
        std::cout << cudaGetErrorString(error) << "\n";                                             \
        exit(1);                                                                                    \
    }                                                                                               \
} while (0)

#define CUSOLVER_ERROR_CHECK(s)                                                                          \
do {                                                                                                        \
    cusolverStatus_t error = s;                                                                          \
    if (error != CUSOLVER_STATUS_SUCCESS) {                                                              \
        std::cout << "cuSOLVER Error " << error << " at " << __FILE__ << ":" << __LINE__ << "\n";        \
        exit(1);                                                                                         \
    }                                                                                                    \
} while (0)
```

bonachea commented on include/sympack/cuBLAS.hpp:
```diff
@@ -0,0 +1,627 @@
+        int gpu_id;
+        int n_gpus;
+
+        MPI_Comm_rank(symPACK::world_comm, &rank);
+        cudaGetDeviceCount(&n_gpus);
+
+        gpu_id = rank % n_gpus;
+
+        cudaSetDevice(gpu_id);
```
IIUC the lines above (and similar lines duplicated in template functions below) are selecting a local GPU based on the _global_ rank of this process. This happens to have reasonable behavior for job layout where tasks-per-node happen to be a multiple of gpus-per-node, but I'm pretty sure it "does the wrong thing" for job layouts lacking this property (e.g. 1 tasks-per-node and 2 GPUs-per-node). 

It would make slightly more sense to use the _local_ rank of the process (ie. `upcxx::local_team().rank_me()`), but would be even better to just use `symPACK::gpu_allocator.device_id()`:

```suggestion
        cudaSetDevice(symPACK::gpu_allocator.device_id());
```

Wider question: `symPACK_cuda_setup()` appears to already do exactly that, so **why are these `cudaSetDevice()` calls required at all?**

Hooninator replied:
See answer to comment below, this is dead code that never should have been in the PR. Sorry for the confusion caused by this.

bonachea commented on include/sympack/cuBLAS.hpp:
```diff
@@ -0,0 +1,627 @@
+        CUDA_ERROR_CHECK(cudaMalloc(reinterpret_cast<void**>(&d_C), ldc * N * sizeof(C[0])));
+        CUDA_ERROR_CHECK(cudaMemcpyAsync(d_C, C, ldc * N * sizeof(C[0]), cudaMemcpyHostToDevice));
+
+        /* Do GEMM */
```
What is the purpose of these `cublas::cublas_FUNC_wrapper` template functions?
IIUC they only perform host/device data movement for the arguments, but don't actually perform the advertised computation! They also appear to be completed unused. This is in contrast to the `cublas::cublas_FUNC_wrapper2` template functions, that _are_ used and appear to simply invoke cuBLAS.

Is this dead code left over from some previous version? 
If they are intended to serve a useful purpose then the comment at the top of the file should probably explain, because right now their existence seems very misleading.

Hooninator replied:
You’re right, this is dead code that should not have been included in the PR. Apologies, I will remove all of this.

Hooninator replied:
Dead code has been removed in [cb5d13b](https://github.com/symPACK/symPACK/pull/12/commits/cb5d13b01d569f7077d6f51f0ea15788200b0a6c)

bonachea commented on include/sympack/symPACKMatrix2D.hpp:
```diff
@@ -651,6 +726,7 @@ namespace symPACK{
         }
 
         // Copy constructor.  
+        // TODO: Edit this with new fields
```
IIUC the `blockCell_t` copy/move constructors and copy/move assignment have not been updated to handle the new CUDA_MODE fields (`_d_nzval`, `is_gpu_block`). I'd expect this to cause potentially subtle problems **if** these are ever invoked in CUDA_MODE, and the short comment above is unlikely to help find the problem.

Are these operations simply unreachable?

If so, then for maintainability reasons I recommend inserting the following in each such function:
```
#ifdef CUDA_MODE
throw std::runtime_error("This function is unimplemented for CUDA_MODE").
#endif
```

Hooninator replied:
These constructors/operators are only called during the Solve phase, which does not currently use the strategy that the Factorize phase uses to bypass the intermediate copy to host memory in certain cases. As a result, it is impossible at any point for a `blockCell_t` instance that these constructors/operators interact with to have either the `_d_nzval` or `is_gpu_block` field set to anything other than `nullptr` and `false`, respectively. 

I will add a more detailed comment to the code explaining that these constructors will need to be edited if someone ever wants to add the capability to bypass host memory to the communication present in the Solve phase.

bonachea replied:
> it is impossible at any point for a `blockCell_t` instance that these constructors/operators interact with to have either the `_d_nzval` or `is_gpu_block` field set to anything other than `nullptr` and `false`, respectively.
> 
> I will add a more detailed comment to the code explaining that these constructors will need to be edited if someone ever wants to add the capability to bypass host memory to the communication present in the Solve phase.

Thanks for the clarification.

You've stated a behavioral invariant, and IMO such invariants are always much more useful when embodied as an assertion rather than a comment.

How about inserting the following in each function:
```
#ifdef CUDA_MODE
UPCXX_ASSERT(!_d_nzval && !is_gpu_block, "This function is unimplemented for blocks using host-bypass communication");
#endif
```

Hooninator replied:
Fixed here [d7d69d0](https://github.com/symPACK/symPACK/pull/12/commits/d7d69d0a4241a680d3940fc9ea807395f75a0548)

bonachea commented on include/sympack/symPACKMatrix2D.hpp:
```diff
@@ -897,16 +1077,59 @@ namespace symPACK{
+            }
+            
+            if (diag_nzval_inter==nullptr || nzblk_nzval_inter==nullptr) {
+                blas::Trsm('L','U','T','N',snode_size, total_rows(), T(1.0),  diag_nzval, snode_size, nzblk_nzval, snode_size);
```
IIUC, this branch handles the case where one or both of the calls to `symPACK::gpu_allocator.allocate` above returns `nullptr` (ie. due to device segment exhaustion).

However, for the case where one allocation succeeds and the other fails, I believe this control path results in _leaking_ one device allocation, which of course increases the chances that later allocations will also fail.

bonachea commented on include/sympack/symPACKMatrix2D.hpp:
```diff
@@ -1005,26 +1228,105 @@ namespace symPACK{
+#ifdef CUDA_MODE
+              if (tgt_width * src_snode_size > symPACK::syrk_limit) {
+            
+                dev_ptr d_pivot_nzval = symPACK::gpu_allocator.allocate<T>(tgt_width * src_snode_size);
+                dev_ptr d_buf = symPACK::gpu_allocator.allocate<T>(tgt_width * ldbuf);
+                
+                if (d_pivot_nzval==nullptr) {
+                    blas::Syrk('U','T',tgt_width, src_snode_size, T(-1.0), pivot_nzval, src_snode_size, beta, buf, ldbuf);
+                } else if (d_buf==nullptr) { 
+                    blas::Syrk('U','T',tgt_width, src_snode_size, T(-1.0), pivot_nzval, src_snode_size, beta, buf, ldbuf);
```
Similar observation here - the first conditional path above can sometimes leak `d_buf`, and the second is guaranteed to leak `d_pivot_nzval`

bonachea commented on include/sympack/symPACKMatrix2D.hpp:
```diff
@@ -1005,26 +1228,105 @@ namespace symPACK{
+              if (tgt_width * src_nrows > symPACK::gemm_limit ||
+                  tgt_width * src_snode_size > symPACK::gemm_limit ||
+                  src_nrows * src_snode_size > symPACK::gemm_limit) {
+                dev_ptr d_pivot_nzval = symPACK::gpu_allocator.allocate<T>(tgt_width * src_snode_size);
+                dev_ptr d_facing_nzval = symPACK::gpu_allocator.allocate<T>(src_nrows * src_snode_size);
+                dev_ptr d_buf = symPACK::gpu_allocator.allocate<T>(ldbuf * src_nrows);
+                if (d_pivot_nzval==nullptr) {
+                    blas::Gemm('T','N',tgt_width, src_nrows,src_snode_size,
+                                T(-1.0),pivot_nzval,src_snode_size,
+                                facing_nzval,src_snode_size,beta,buf,ldbuf);
+                } else if (d_facing_nzval==nullptr) {
+                    blas::Gemm('T','N',tgt_width, src_nrows,src_snode_size,
+                                T(-1.0),pivot_nzval,src_snode_size,
+                                facing_nzval,src_snode_size,beta,buf,ldbuf);
+                } else if (d_buf==nullptr) { 
+                    blas::Gemm('T','N',tgt_width, src_nrows,src_snode_size,
+                                T(-1.0),pivot_nzval,src_snode_size,
+                                facing_nzval,src_snode_size,beta,buf,ldbuf);
```
More leaks here.

Note that `device_allocator::deallocate(nullptr)` is defined as a no-op, so the most elegant solution in this particular case is probably to ensure all three fallback paths (which all seem to take the same action?) all reach the three deallocation calls below (ie by moving them past the enclosing `}`) and let the allocator sort it out.

bonachea commented on include/sympack/symPACKMatrix2D.hpp:
```diff
@@ -1163,11 +1455,37 @@ namespace symPACK{
+#ifdef CUDA_MODE
+            if (ldsol*ldfact > symPACK::trsm_limit && symPACK::gpu_solve) {
+            
+                dev_ptr d_diag_nzval = symPACK::gpu_allocator.allocate<T>(ldfact*ldfact);
+                dev_ptr d_nzval = symPACK::gpu_allocator.allocate<T>(ldsol*ldfact);
```
These device allocations (and four more similar blocks below) are not checking whether the allocation succeeded. We don't necessarily need a fallback implementation to handle every such allocation failure, but currently the behavior is likely to be an unfriendly CUDA error about null pointers inside the `upcxx::copy` operation below.

I'd prefer an assertion to provide a clear diagnosis of what went wrong. Eg:

```suggestion
                dev_ptr d_diag_nzval = symPACK::gpu_allocator.allocate<T>(ldfact*ldfact);
                dev_ptr d_nzval = symPACK::gpu_allocator.allocate<T>(ldsol*ldfact);
                UPCXX_ASSERT_ALWAYS(d_diag_nzval && d_nzval,
                  "GPU device segment exhausted, try increasing `-gpu_mem` (see README.md)");
```

If you're feeling fancy you could even factor the message and include the values of `symPACK::gpu_allocator.segment_size()` and `symPACK::gpu_allocator.segment_used()` in the error message, but such queries should be conditional on `#if UPCXX_SPEC_VERSION >= 20230300`

Hooninator replied:
Thanks for catching these memory leaks. I will implement and push a fix.

Hooninator replied:
Fixed in [97a3e5f](https://github.com/symPACK/symPACK/pull/12/commits/97a3e5f1f6d37fa0e56f7c6290ef8c9722114254)

bonachea commented on include/sympack/symPACKMatrix2D.hpp:
```diff
@@ -1005,26 +1228,105 @@ namespace symPACK{
+                    symPACK::gpu_allocator.deallocate(d_buf);
+                }
 
+              } else {
+                 blas::Gemm('T','N',tgt_width, src_nrows,src_snode_size,
+                            T(-1.0),pivot_nzval,src_snode_size,
+                            facing_nzval,src_snode_size,beta,buf,ldbuf);
+              }
+#else
               blas::Gemm('T','N',tgt_width, src_nrows,src_snode_size,
                   T(-1.0),pivot_nzval,src_snode_size,
                   facing_nzval,src_snode_size,beta,buf,ldbuf);
+#endif                  
```
This is purely a stylistic observation, but I believe this pattern (which is repeated multiple times in this file) could more concisely be written as:

```suggestion
              } else 
#endif
              blas::Gemm('T','N',tgt_width, src_nrows,src_snode_size,
                  T(-1.0),pivot_nzval,src_snode_size,
                  facing_nzval,src_snode_size,beta,buf,ldbuf);     
```

That way you have one fewer copy of the code making the complicated blas call, which helps with maintainability.

A possibly superior approach is to put the complicated blas call in a zero-argument lambda expression assigned to a variable at the entry to this logic, and then invoke that lambda for all the "use the CPU" cases (including the device allocation failure cases).

Hooninator replied:
Fixed in [97a3e5f](https://github.com/symPACK/symPACK/pull/12/commits/97a3e5f1f6d37fa0e56f7c6290ef8c9722114254)

bonachea commented on src/global.cpp:
```diff
@@ -36,27 +36,82 @@ int symPACK_Rank(int * rank){
+  
+  size_t alloc_size, free, total;
+  CUDA_ERROR_CHECK(cudaMemGetInfo(&free, &total));
+  alloc_size = (free/2) / (std::max(tasks_per_node, n_gpus) / n_gpus);
```
On a related note, why does the expression include `free/2`?
I agree with the idea that we should not attempt to allocate every last byte of available GPU memory as reported by `cudaMemGetInfo`, but discarding half the available space seems excessive. Is the balance of this GPU memory being reserved for some purpose outside the UPC++ device segment, or can we safely increase this to a larger fraction, eg:

```suggestion
  alloc_size = (size_t)(free*0.9) / (std::max(tasks_per_node, n_gpus) / n_gpus);
```

bonachea commented on README.md:
@Hooninator the README documents toolchain files in `toolchains/*.cmake`, and specifically mentions NERSC Cori and cori.cmake (which is presumably obsolete and should probably be removed).

Have you been using a toolchain file for Perlmutter that is worth sharing?
E.g. did you install your own Scotch and METIS or find a NERSC-supported install somewhere?

Hooninator replied:
I haven't been using a toolchain file that would be worth sharing, so I'll just remove the reference to the cori toolchain from the documentation.

Hooninator replied:
fixed [5b584c8](https://github.com/symPACK/symPACK/pull/12/commits/5b584c855c2a0e81ae93a70f7aa9a5a361ae3bde)

bonachea commented on README.md:
````diff
@@ -74,7 +80,7 @@ make
 ```
 
-Additionally, standalone drivers for **symPACK** can be built by typing `make examples`
+Additionally, a standalone driver for **symPACK** can be built by typing `make run_sympack2D`.
````
Is there any good reason why the `run_sympack2D` driver is not built by the default make target?

Hooninator replied:
I guess there’s no reason why it shouldn’t be built by default. I’ll fiddle with this.

Hooninator replied:
fixed in [3b0ddd3](https://github.com/symPACK/symPACK/pull/12/commits/3b0ddd3f00759fe368d22285bb33cd3e23a259a9)

bonachea commented on README.md:
```diff
@@ -118,12 +128,22 @@ wget https://sparse.tamu.edu/RB/Nasa/nasa2146.tar.gz
+- `-gpu_mem <SIZE>` controls the amount of memory (in bytes) allocated on the GPU upon calling the `symPACK_cuda_setup()` method. If this option is not specified, **symPACK** will attempt partition a device's memory equally among each process bound to the device. For job configurations that have more than one GPU bound to each node, this default behavior may under-estimate the amount of memory th ...
+- `-gpu_blk <SIZE>` controls the size threshold (in bytes) that a factorized diagonal block must exceed for it to be automatically copied to the GPU bound to a remote process when being sent to said remote process. This allows for an intermediate copy to host memory to be bypassed, which can reduce communication overhead for certain problems.
+- `{trsm, gemm, potrf, syrk}_limit <SIZE>` controls how many nonzero entries a block must have for the given matrix operation (TRSM, GEMM, POTRF, or SYRK) involving said block to be offloaded to the GPU. The GPU will generally be most beneficial for operations involving large blocks with many nonzeros, but the exact size thresholds for when each operation starts to run faster on the GPU will diff ...
+- `-gpu_solve` determines if the GPU will be used for **symPACK's** triangular solve routine.
+- `-fallback {cpu, terminate}` determines what **symPACK** will do in the event that an attempt to allocate memory on a device for use in a matrix computation fails due to an insufficient amount of free memory. Setting this option to `cpu` will make it so **symPACK** will simply perform the computation on the CPU instead of the GPU. Setting this option to `terminate` will cause **symPACK** to thr ...
```
Would like us to say more about the default behavior for command-line arguments.

I suggest at least:

```suggestion
- `-gpu_solve` determines if the GPU will be used for **symPACK's** triangular solve routine. Defaults to off (use the CPU for the triangular solve).
- `-fallback {cpu, terminate}` determines what **symPACK** will do in the event that an attempt to allocate memory on a device for use in a matrix computation fails due to an insufficient amount of free memory. Setting this option to `cpu` will make it so **symPACK** will simply perform the computation on the CPU instead of the GPU. Setting this option to `terminate` will cause **symPACK** to throw an error and cease execution when a device allocation fails. Defaults to `terminate`
```


I'm less sure we should encode the `-gpu_mem` and `-gpu_blk` numeric defaults in this document. Ideally we'd have a `-v` option to enable verbosity that reports the active values of all these voodoo tuning parameters. 


Thoughts?

Hooninator replied:
Yeah the default values of each of these definitely need to be specified in the README. It would be easy and helpful to add a verbosity option, so I will do so.

Hooninator replied:
verbosity option added in commit [f6d8093] (https://github.com/symPACK/symPACK/pull/12/commits/f6d809341bcba5312b9f1752663877df14e58e6e)

bonachea commented on include/sympack/symPACKMatrix2D.hpp:
```diff
@@ -651,24 +726,31 @@ namespace symPACK{
+        // TODO: Edit this with new fields
         blockCell_t ( const blockCell_t & other ): blockCell_t() {
+#ifdef CUDA_MODE
+          UPCXX_ASSERT(!_d_nzval && !is_gpu_block, "This function is unimplemented for blocks using host-bypass communication");
```
Sorry, I of course meant all these assertions to be on the source object, sorry to inadvertently mislead you.
```suggestion
          UPCXX_ASSERT(!other._d_nzval && !other.is_gpu_block, "This function is unimplemented for blocks using host-bypass communication");
```

Hooninator replied:
fixed in [26f2b7d](https://github.com/symPACK/symPACK/pull/12/commits/26f2b7d9f549cb2b1a6fe3a17af42076c047452f)

bonachea commented on README.md:
```diff
@@ -54,14 +55,14 @@ The `...OPTIONS...` can be one of the following:
+
 ### Notes for specific platforms
 
 Some platforms have preconfigured toolchain files which can be used by adding the following option to the `cmake` command:
```
The line above no longer makes sense with the example below removed entirely.

How about replacing the example with one that invokes a hypothetical `mysystem.cmake` file?

Incidentally, we should also remove the obsolete cori.cmake file.

Hooninator replied:
fixed in [7551c40](https://github.com/symPACK/symPACK/pull/12/commits/7551c406acf2fadb6666aa41ccc2112ceb78f49f)

bonachea commented on src/global.cpp:
```diff
@@ -36,27 +36,99 @@ int symPACK_Rank(int * rank){
+  if (optionsFact.gpu_verbose) {
+    if (upcxx::rank_me()==0) {
+        std::cout<<"========= GPU CONFIGURATION OPTIONS ========="<<std::endl;
+        std::cout<<"-gpu_mem: "<<symPACK::gpu_alloc_size<<" bytes"<<std::endl;
```
Weak request to report in MiB for readability:
```suggestion
        std::cout<<"-gpu_mem: "<<(symPACK::gpu_alloc_size/(double)(1<<20))<<" MiB"<<std::endl;
```
would be even cooler if the `-gpu_mem` command line arg accepted an "M*" or "G*" suffix

Hooninator replied:
optional suffix added in [119993d](https://github.com/symPACK/symPACK/pull/12/commits/119993d975ffe5ae5a166d3fad5f72f84f1de936)

bonachea commented on driver/utils.hpp:
```diff
@@ -221,6 +221,61 @@ inline void process_options(int argc, char **argv, symPACK::symPACKOptions & opt
+ } 
+ 
+ optionsFact.gpu_verbose = false;
+ if (options.find("-gpu_v") != options.end()) {
```
This option is very helpful, please document in README.md

Hooninator replied:
fixed in [d0bb0dc](https://github.com/symPACK/symPACK/pull/12/commits/d0bb0dcb05d162eedffdfbdd04d683f3dce33230)

bonachea commented:
@EGNg @mathias-cerebras @PHHargrove as you probably know there is a soft deadline (which I won't describe here) for this work happening today.

If you intend to provide any more feedback on this PR, it would be most useful today.

bonachea commented on driver/utils.hpp:
```diff
@@ -196,7 +195,23 @@ void check_solution( int n, vdist_int * vertexDist, ptr_t * colptr, ind_t * rowi
+    else if (alloc_units=="GiB")
+        size *= (1<<30);
+    
+    return size;
```
Thanks for adding this nice usability improvement!

However I'm concerned the implementation above doesn't contain any "guardrails" to validate the user input is correctly recognized, leading to potential nonsense. In particular, I'm concerned about silent misbehavior if the user types one of the following:

```
run_sympack2D -gpu_mem 2 GiB
run_sympack2D -gpu_mem 2.5GB
run_sympack2D -gpu_mem 2GB
run_sympack2D -gpu_mem 2G
run_sympack2D -gpu_mem 2gib
run_sympack2D -gpu_mem 2
```

These all technically violate the documentation in README, but you can understand how a user might easily do these. I suggest the following enhancements:

1. Append all of `options["-gpu_mem"]` to include trailing options
2. Parse the number as floating point
3. Interpret any unit that starts with [gGmM] as GiB or MiB respectively and ignore subsequent characters
4. Issue an error if the first unit character is anything else
5. Issue an error if the parsed gpu_mem segment value is less than 1 MiB (which is effectively nonsense, and means we misunderstood the user)

If you're feeling ambitious, you could even factor this parsing function, add a [kK] suffix for KiB, and use it to also parse memory size value provided in the `-gpu_blk` and `-*_limit` arguments (although I think those are far less likely to be tweaked in practice by end users).

bonachea replied:
Resolved as suggested in c62c5ad

bonachea commented on include/sympack/symPACKMatrix2D.hpp:
```diff
@@ -865,30 +969,149 @@ namespace symPACK{
+                }
+                if (symPACK::gpu_verbose) {
+                    symPACK::cpu_ops["potrf"]+=1;
+                } 
```
@Hooninator Thanks for adding these "rough" counters. 

Some suggestions for improvement:
1. Move the CPU counter increments into the lambda above that invokes the operation
2. Move the GPU counter increments into the "delay slot" between the lapack/cublas invocation and the cudaDeviceSynchronize() that finishes the operation
3. Create an inline function to factor, eg:
```
// factored definition in Environment.hpp
template <typename Map>
inline void symPACK::counter(Map &map, std::string const &op) {
  if (symPACK::gpu_verbose) map[op]++;
}
// use examples:
symPACK::counter(symPACK::gpu_ops, "potrf");
symPACK::counter(symPACK::cpu_ops, "gemm");
```

Hooninator replied:
Fixed in [4d62638](https://github.com/symPACK/symPACK/pull/12/commits/4d62638fe283eff6a4cea951deb5883d520bb1e6)

bonachea commented on include/sympack/Environment.hpp:
```diff
@@ -80,8 +98,12 @@ namespace symPACK{
-
+  template <typename Map>
+  inline void increment_counter(Map &map, std::string const &op) {
+    if (symPACK::gpu_verbose) {
+        map[op]+=1;
+    }
```
@Hooninator This is now failing to compile in non-CUDA mode, because `symPACK::gpu_verbose` and `gpu_ops/cuda_ops` require CUDA_MODE

I suggest:

```suggestion
  #ifdef CUDA_MODE
    if (symPACK::gpu_verbose) {
        map[op]+=1;
    }
  #endif
```
And also hoisting `cpu_ops, gpu_ops` definitions out of `CUDA_MODE` to avoid compile errors (although they will be unused in non-CUDA mode)

Hooninator replied:
Whoops, should have caught this. Thank you.

mathias-cerebras commented on scripts/analyze_time.py:
@Hooninator is this script described in a readme? Otherwise it would be great to have a one liner showing how it's used.

Thanks a lot for all the changes

Hooninator replied:
That script was something I used a few times early in the summer and is not meant to be used by anyone else. I meant to remove it a while ago. This is now fixed in [f7fc016](https://github.com/symPACK/symPACK/pull/12/commits/f7fc0164197cddecf012bd706c8ca005f197ebb7)

bonachea commented on src/global.cpp:
```diff
@@ -36,27 +36,108 @@ int symPACK_Rank(int * rank){
-  // init MPI, if necessary
-  MPI_Initialized(&symPACK::mpi_already_init);
-  if (!symPACK::mpi_already_init) MPI_Init(argc, argv);
-  if ( symPACK::world_comm == MPI_COMM_NULL ) MPI_Comm_split(MPI_COMM_WORLD, 0, upcxx::rank_me(), &symPACK::world_comm);
```
@Hooninator I just noticed this change, converting symPACK to init MPI before UPC++.

We support either order on platforms like the HPE Cray EX, although there can be some subtle interactions on other platforms so all else being equal [we usually recommend UPC++ be init first](https://bitbucket.org/berkeleylab/upcxx/wiki/docs/mpi-hybrid.md).

Do you recall why this change was made? 

Regardless of init order, the final line above that renumbers MPI_COMM_WORLD to ensure symPACK::world_comm matches the rank ordering of UPC++ should remain where it was (after both libraries are init) and unchanged (otherwise it defeats the purpose).

bonachea replied:
After further investigation, I believe this was probably changed to workaround [bug 4638](https://upc-bugs.lbl.gov/bugzilla/show_bug.cgi?id=4638), which is believed to arise from a defect in Cray MPI that is deployed on HPE Cray EX Slingshot-11 systems.

There are other (envvar-based) workarounds available for that bug. Applying the symPACK-level source change has the advantage of more user-friendly behavior on Slingshot systems (no user-driven workaround needed), but may potentially complicate/prevent spawning on less integrated systems (e.g. using GASNet ssh-spawner). 

On balance I think I'm inclined to leave the init order proposed in this PR (MPI then GASNet) at least until bug 4638 has a better resolution, although we might revisit that decision in the future if it creates too many problems.

In any case I've just pushed 2838378 to repair the MPI rank renumbering regression mentioned above.

CC: @PHHargrove

bonachea commented on include/sympack/symPACKMatrix2D.hpp:
```diff
@@ -5,8 +5,18 @@
+#include "cusolverDn.h"
+#endif
+
+#include <omp.h>
```
This include appears to be have added for no reason (and unconditionally), and runs at least a small risk of introducing a conflict with some OpenMP implementations.

```suggestion
```

bonachea commented on include/sympack/symPACKMatrix2D.hpp:
```diff
@@ -22,15 +32,13 @@
 #define _PRIORITY_QUEUE_RDY_
 
 
-//#define _USE_PROM_AVAIL_
-//#define _USE_PROM_RDY_
+#define _USE_PROM_AVAIL_
+#define _USE_PROM_RDY_
```
These settings are a behavioral change for non-CUDA mode relative to current master branch.

I don't know any reason to object to this change, just wanted to note it for completeness.

bonachea reviewed (changes requested):
Added a few more comments after more detailed review.

bonachea commented:
This PR has caused a behavioral regression for `-DENABLE_THREADS=ON` mode.

Previously one could `cmake -DENABLE_THREADS=ON ... && make` to build threaded symPACK (or `make all examples` to build the threaded 1-D solver) and everything works.

Now in `-DENABLE_THREADS=ON` mode, the 1-D solver continues to build if you request it by name (`make run_sympack`), but the 2-D solver that is built by `make` generates a compile error:
```
.../sympack/include/sympack/symPACKMatrix2D.hpp:6930:29: error: 'class symPACK::scheduling::Scheduler2D<symPACK::scheduling::task_t<std::tuple<unsigned int, unsigned int, symPACK::Factorization::op_type, unsigned int, unsigned int>, std::tuple<int, int> >, symPACK::scheduling::task_graph_t3<symPACK::scheduling::key_t, symPACK::scheduling::task_t<std::tuple<unsigned int, unsigned int, symPACK::Factorization::op_type, unsigned int, unsigned int>, std::tuple<int, int> > > >' has no member named 'scheduler'; did you mean 'Scheduler2D'?
 6930 |                   if (this->scheduler.extraTaskHandle_!=nullptr) {
      |                       ~~~~~~^~~~~~~~~
      |                       Scheduler2D

```

I suspect the 2-D + THREADS combo has probably never worked, but the fact that this PR introduces building the 2-D solver by default means the `-DENABLE_THREADS=ON` option is now effectively broken. I see several possible resolutions: (ranked from least to most estimated effort)

1. Clearly document `-DENABLE_THREADS=ON` as "experimental" and state it's only supported for the deprecated 1-D solver, built via `make run_sympack`
2. Change CMake logic so we don't attempt to build the 2-d solver in `-DENABLE_THREADS=ON` mode unless specifically requested
3. Try to repair the 2-D + THREADS combo

bonachea commented:
2838378 rebased onto master resulting to 804c30f , which is textually identical (aside from a trivial whitespace difference)

This was done in order to establish a linear git history. Please do not push any more merge commits.

bonachea reviewed (changes requested):
This PR has caused a behavioral regression for `-DENABLE_THREADS=ON` mode.

Previously (on the master branch) one could `cmake -DENABLE_THREADS=ON ... && make` to build threaded symPACK (or `make examples` to build the threaded 1-D and 2-D solvers), all of which compiled without errors. Both the 1-D and 2-D solvers build from master appear to work correctly at runtime with threading enabled.

Now with the PR in `-DENABLE_THREADS=ON` mode, the 1-D solver continues to build if you request it by name (`make run_sympack`), but the 2-D solver that is built by `make` generates a compile error:
```
.../sympack/include/sympack/symPACKMatrix2D.hpp:6930:29: error: 'class symPACK::scheduling::Scheduler2D<symPACK::scheduling::task_t<std::tuple<unsigned int, unsigned int, symPACK::Factorization::op_type, unsigned int, unsigned int>, std::tuple<int, int> >, symPACK::scheduling::task_graph_t3<symPACK::scheduling::key_t, symPACK::scheduling::task_t<std::tuple<unsigned int, unsigned int, symPACK::Factorization::op_type, unsigned int, unsigned int>, std::tuple<int, int> > > >' has no member named 'scheduler'; did you mean 'Scheduler2D'?
 6930 |                   if (this->scheduler.extraTaskHandle_!=nullptr) {
      |                       ~~~~~~^~~~~~~~~
      |                       Scheduler2D

```

I'm not sure of the 2-D + THREADS combo was ever recommended, but the fact that this PR introduces building the 2-D solver by default means the `-DENABLE_THREADS=ON` option defaults to a broken build. 

I see several potential resolutions: (ranked from least to most estimated effort)

1. Clearly document `-DENABLE_THREADS=ON` as "experimental" and state it's only supported for the deprecated 1-D solver, built via `make run_sympack`
2. Change CMake logic so we don't attempt to build the 2-d solver in `-DENABLE_THREADS=ON` mode unless explicitly requested
3. Try to repair the 2-D + THREADS combo

@Hooninator I don't expect you to solve this problem, but your thoughts on how to resolve this regression are appreciated.

mathias-cerebras commented:
> This PR has caused a behavioral regression for `-DENABLE_THREADS=ON` mode.
> 
> Previously (on the master branch) one could `cmake -DENABLE_THREADS=ON ... && make` to build threaded symPACK (or `make examples` to build the threaded 1-D and 2-D solvers), all of which compiled without errors. Both the 1-D and 2-D solvers build from master appear to work correctly at runtime with threading enabled.
> 
> Now with the PR in `-DENABLE_THREADS=ON` mode, the 1-D solver continues to build if you request it by name (`make run_sympack`), but the 2-D solver that is built by `make` generates a compile error:
> 
> ```
> .../sympack/include/sympack/symPACKMatrix2D.hpp:6930:29: error: 'class symPACK::scheduling::Scheduler2D<symPACK::scheduling::task_t<std::tuple<unsigned int, unsigned int, symPACK::Factorization::op_type, unsigned int, unsigned int>, std::tuple<int, int> >, symPACK::scheduling::task_graph_t3<symPACK::scheduling::key_t, symPACK::scheduling::task_t<std::tuple<unsigned int, unsigned int, symPACK::Factorization::op_type, unsigned int, unsigned int>, std::tuple<int, int> > > >' has no member named 'scheduler'; did you mean 'Scheduler2D'?
>  6930 |                   if (this->scheduler.extraTaskHandle_!=nullptr) {
>       |                       ~~~~~~^~~~~~~~~
>       |                       Scheduler2D
> ```
> 
> I'm not sure of the 2-D + THREADS combo was ever recommended, but the fact that this PR introduces building the 2-D solver by default means the `-DENABLE_THREADS=ON` option defaults to a broken build.
> 
> I see several potential resolutions: (ranked from least to most estimated effort)
> 
> 1. Clearly document `-DENABLE_THREADS=ON` as "experimental" and state it's only supported for the deprecated 1-D solver, built via `make run_sympack`
> 2. Change CMake logic so we don't attempt to build the 2-d solver in `-DENABLE_THREADS=ON` mode unless explicitly requested
> 3. Try to repair the 2-D + THREADS combo
> 
> @Hooninator I don't expect you to solve this problem, but your thoughts on how to resolve this regression are appreciated.

I believe the `this` pointer captured by the lambda is the `Scheduler2D` object, so the correct code should be 
```
#ifdef SP_THREADS
                  if (this->extraTaskHandle_!=nullptr) {
                  bool delay = this->extraTaskHandle_(ptr);
                  if (delay) {
                  this->delayedTasks_[ptr->_lock_ptr].push_back(ptr);
                  return;
                  }
                  }
#endif
```

Hooninator commented:
> This PR has caused a behavioral regression for `-DENABLE_THREADS=ON` mode.
> 
> Previously (on the master branch) one could `cmake -DENABLE_THREADS=ON ... && make` to build threaded symPACK (or `make examples` to build the threaded 1-D and 2-D solvers), all of which compiled without errors. Both the 1-D and 2-D solvers build from master appear to work correctly at runtime with threading enabled.
> 
> Now with the PR in `-DENABLE_THREADS=ON` mode, the 1-D solver continues to build if you request it by name (`make run_sympack`), but the 2-D solver that is built by `make` generates a compile error:
> 
> ```
> .../sympack/include/sympack/symPACKMatrix2D.hpp:6930:29: error: 'class symPACK::scheduling::Scheduler2D<symPACK::scheduling::task_t<std::tuple<unsigned int, unsigned int, symPACK::Factorization::op_type, unsigned int, unsigned int>, std::tuple<int, int> >, symPACK::scheduling::task_graph_t3<symPACK::scheduling::key_t, symPACK::scheduling::task_t<std::tuple<unsigned int, unsigned int, symPACK::Factorization::op_type, unsigned int, unsigned int>, std::tuple<int, int> > > >' has no member named 'scheduler'; did you mean 'Scheduler2D'?
>  6930 |                   if (this->scheduler.extraTaskHandle_!=nullptr) {
>       |                       ~~~~~~^~~~~~~~~
>       |                       Scheduler2D
> ```
> 
> I'm not sure of the 2-D + THREADS combo was ever recommended, but the fact that this PR introduces building the 2-D solver by default means the `-DENABLE_THREADS=ON` option defaults to a broken build.
> 
> I see several potential resolutions: (ranked from least to most estimated effort)
> 
> 1. Clearly document `-DENABLE_THREADS=ON` as "experimental" and state it's only supported for the deprecated 1-D solver, built via `make run_sympack`
> 2. Change CMake logic so we don't attempt to build the 2-d solver in `-DENABLE_THREADS=ON` mode unless explicitly requested
> 3. Try to repair the 2-D + THREADS combo
> 
> @Hooninator I don't expect you to solve this problem, but your thoughts on how to resolve this regression are appreciated.

Fixed in [fb329af](https://github.com/symPACK/symPACK/pull/12/commits/fb329af80a133d8fa9a0a6452e31b235b71fef8c)

Basically, the PR enables a build variable called `_USE_PROM_RDY_` which was de-activated by default previously. Enabling `_USE_PROM_RDY_` in conjunction with `DENABLE_THREADS=ON` forced the compiler to try to compile certain parts of the code that it previously did not touch. These regions of the code were in fact incorrectly written from the get-go in the current master branch, but since the current master branch has `_USE_PROM_RDY_` de-activated, the errors went unnoticed.

bonachea reviewed (approved):
After much iteration, I'm finally satisfied with this PR as it stands at c62c5ad

@mathias-cerebras @EGNg please speak up ASAP if you want more time to review this PR before it is merged.

mathias-cerebras commented:
On my end things look good (from what I understood of the PR). Thanks a lot @bonachea and @Hooninator

Hooninator commented:
> On my end things look good (from what I understood of the PR). Thanks a lot @bonachea and @Hooninator

@mathias-cerebras Thanks for looking this over! You're listed as a reviewer for this PR, so if everything looks good to you, would you mind approving the PR?

mjacquel reviewed (approved):
Thanks @Hooninator

Hooninator commented:
@mjacquel Sorry to be a pest about this, but it looks like your cerebras account (@mathias-cerebras) is also listed as a reviewer, so I think you need to also approve the PR with that account. So sorry for the inconvenience!

Hooninator commented:
@bonachea @mjacquel would one of you be able to merge this PR when you have a moment? I don't have write access to the symPACK repo, so I cannot do it. 

Thank you in advance.

PHHargrove commented:
> @bonachea @mjacquel would one of you be able to merge this PR when you have a moment? I don't have write access to the symPACK repo, so I cannot do it.
> 
> Thank you in advance.

@Hooninator, right now @bonachea is on vacation.
I am pretty sure merging this PR is on the list of things he'll be doing sometime after he returns.

Hooninator commented:
> > @bonachea @mjacquel would one of you be able to merge this PR when you have a moment? I don't have write access to the symPACK repo, so I cannot do it.
> > Thank you in advance.
> 
> @Hooninator, right now @bonachea is on vacation. I am pretty sure merging this PR is on the list of things he'll be doing sometime after he returns.

Oh shoot, my bad, I didn't realize he was on vacation! Thank you for letting me know.
